src/abstractdataframe/join.jl

```diff
@@ -81,16 +81,350 @@ end
 
 Base.length(x::RowIndexMap) = length(x.orig)
 
-# composes the joined data table using the maps between the left and right
-# table rows and the indices of rows in the result
-
 _rename_cols(old_names::AbstractVector{Symbol},
              renamecols::Union{Function, Symbol, AbstractString},
              exclude::AbstractVector{Symbol} = Symbol[]) =
     Symbol[n in exclude ? n :
            (renamecols isa Function ? Symbol(renamecols(string(n))) : Symbol(n, renamecols))
            for n in old_names]
 
+prepare_on_col() = throw(ArgumentError("at least one on column required when joining"))
+prepare_on_col(c::AbstractVector) = c
+prepare_on_col(cs::AbstractVector...) = tuple.(cs...)
+
+# Return if it is allowed to use refpool instead of the original array for joining.
+# There are multiple conditions that must be met to allow for this.
+# If it is allowed we are sure that nothing can be used as a sentinel
+check_mapping_allowed(short::AbstractVector, refarray_long::AbstractVector,
+                      refpool_long, invrefpool_long) =
+    !isempty(short) && !isnothing(refpool_long) && !isnothing(invrefpool_long) &&
+        eltype(refarray_long) <: Union{Signed, Unsigned}
+
+@noinline map_refarray(mapping::AbstractVector, refarray::AbstractVector, ::Val{true}) =
+    [@inbounds mapping[r + 1] for r in refarray]
+
+@noinline map_refarray(mapping::AbstractVector, refarray::AbstractVector, ::Val{false}) =
+    [@inbounds mapping[r] for r in refarray]
```

Comment:
Here's what I have in mind to limit duplication:
```suggestion
@noinline map_refarray(mapping::AbstractVector, refarray::AbstractVector, ::Val{fi})  where {fi} =
    [@inbounds mapping[r - fi + 1] for r in refarray]
```